src/OpenTelemetry.Contrib.Instrumentation.Wcf/TelemetryClientMessageInspector.cs

```diff
@@ -0,0 +1,176 @@
+﻿// <copyright file="TelemetryClientMessageInspector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.ServiceModel.Dispatcher;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Contrib.Instrumentation.Wcf.Implementation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Wcf
+{
+    /// <summary>
+    /// An <see cref="IClientMessageInspector"/> implementation which adds telemetry to outgoing requests.
+    /// </summary>
+    public class TelemetryClientMessageInspector : IClientMessageInspector
+    {
+        private static readonly ConcurrentDictionary<string, ActionMetadata> ActionMetadataCache = new ConcurrentDictionary<string, ActionMetadata>();
```

Comment:
ConcurrentDictionary is often slower than using a regular dictionary with a lock. Depending on the cache miss rate expected it might be worth switching to `Hashtable`. `Hashtable` has an interesting [thread safety guarantee](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable?view=net-5.0#thread-safety) which makes it ideal for caches. Basically you only take a lock on writes, you don't need to lock on reads.

Reply:
Thanks for the tip, I'll update.